pkg/kubelet/cm/cpumanager/state/state_file.go

```diff
@@ -0,0 +1,175 @@
+func NewFileState(filePath string) State {
+	stateFile := &stateFile{
+		stateFilePath: filePath,
+		State:         NewMemoryState(),
+	}
+	stateFile.tryRestoreState()
+	return stateFile
+}
+
+// tryRestoreState tries to read state file, upon any error,
+// err message is logged and state is left clean. un-initialized
+func (sf *stateFile) tryRestoreState() {
+	sf.Lock()
+	defer sf.Unlock()
+	var err error
+	invalidStateFile := false
+
+	// used when all parsing is ok
+	tmpAssignments := make(ContainerCPUAssignments)
+	tmpDefaultCPUSet := cpuset.NewCPUSet()
+	tmpContainerCPUSet := cpuset.NewCPUSet()
+
+	var content []byte
+
+	if content, err = ioutil.ReadFile(sf.stateFilePath); os.IsNotExist(err) {
+		// Create file
+		if _, err = os.Create(sf.stateFilePath); err != nil {
+			glog.Errorf("[cpumanager] state file: unable to create state file \"%s\":%s", sf.stateFilePath, err.Error())
+			panic("[cpumanager] state file not created")
+		}
+		glog.Infof("[cpumanager] state file: created empty state file \"%s\"", sf.stateFilePath)
+	} else {
+		// File exists - try to read
+		var readState stateFileData
+
+		if err = json.Unmarshal(content, &readState); err != nil {
+			glog.Warningf("[cpumanager] state file: could not unmarshal, corrupted state file - \"%s\"", sf.stateFilePath)
+			invalidStateFile = true
+		}
+
+		if !invalidStateFile {
+			if tmpDefaultCPUSet, err = cpuset.Parse(readState.DefaultCPUSet); err != nil {
+				glog.Warningf("[cpumanager] state file: could not parse state file - [defaultCpuSet:\"%s\"]", readState.DefaultCPUSet)
+				invalidStateFile = true
+			}
+		}
+
+		if !invalidStateFile {
+			for containerID, cpuString := range readState.Entries {
+				if tmpContainerCPUSet, err = cpuset.Parse(cpuString); err != nil {
+					glog.Warningf("[cpumanager] state file: could not parse state file - container id: %s, cpuset: \"%s\"", containerID, cpuString)
+					invalidStateFile = true
+				}
+				tmpAssignments[containerID] = tmpContainerCPUSet
+			}
+		}
+
+		if invalidStateFile {
+			sf.State.ClearState()
+			sf.storeState()
+		} else {
+			sf.State.SetDefaultCPUSet(tmpDefaultCPUSet)
+			sf.State.SetCPUAssignments(tmpAssignments)
+
+			glog.V(2).Infof("[cpumanager] state file: restored state from state file \"%s\"", sf.stateFilePath)
+			glog.V(2).Infof("[cpumanager] state file: defaultCPUSet: %s", tmpDefaultCPUSet.String())
+		}
+	}
+	return
+}
+
+// saves state to a file, caller is responsible for locking
+func (sf *stateFile) storeState() error {
+	var content []byte
+	var err error
+
+	writeState := stateFileData{
+		DefaultCPUSet: sf.State.GetDefaultCPUSet().String(),
+		Entries:       map[string]string{},
+	}
+
+	for containerID, cset := range sf.State.GetCPUAssignments() {
+		writeState.Entries[containerID] = cset.String()
+	}
+
+	if content, err = json.Marshal(writeState); err != nil {
+		glog.Errorf("[cpumanager] state file: could not parse state to json")
+		return err
```

Comment:
Why does this only log, in comparison to the write failure below where we panic? The end result is the same (fail to write the file.)

Reply:
:+1: